File: vercel-ai/tour-of-agents/src/sub-workflow-agent.ts
import * as restate from "@restatedev/restate-sdk";
import { openai } from "@ai-sdk/openai";
import { generateText, tool, wrapLanguageModel, Output, stepCountIs } from "ai";
import {
  ClaimPrompt,
  ClaimPromptSchema,
  InsuranceClaim,
  InsuranceClaimSchema,
} from "./utils/types";
import { requestHumanReview } from "./utils/utils";
import { durableCalls } from "@restatedev/vercel-ai-middleware";
const schema = restate.serde.schema;

const run = async (ctx: restate.Context, { prompt }: ClaimPrompt) => {
  const model = wrapLanguageModel({
    model: openai("gpt-4o"),
    middleware: durableCalls(ctx, { maxRetryAttempts: 3 }),
  });

  const { text } = await generateText({
    model,
    system:
      "You are an insurance claim evaluation agent. Use these rules: " +
      "* if the amount is more than 1000, ask for human approval, " +
      "* if the amount is less than 1000, decide by yourself",
    prompt,
    tools: {
      // <start_here>
      humanApproval: tool({
        description: "Ask for human approval for high-value claims.",
        inputSchema: InsuranceClaimSchema,
        execute: async (claim: InsuranceClaim) =>
          ctx.serviceClient(humanApprovalWorfklow).requestApproval(claim),
      }),
      // <end_here>
    },
    stopWhen: [stepCountIs(5)],
    providerOptions: { openai: { parallelToolCalls: false } },
  });
  return text;
};

export const agent = restate.service({
  name: "SubWorkflowClaimApprovalAgent",
  handlers: {
    run: restate.createServiceHandler(
      { input: schema(ClaimPromptSchema) },
      run,
    ),
  },
});

// <start_wf>
export const humanApprovalWorfklow = restate.service({
  name: "HumanApprovalWorkflow",
  handlers: {
    requestApproval: async (ctx: restate.Context, claim: InsuranceClaim) => {
      const approval = ctx.awakeable<boolean>();
      await ctx.run("request-review", () =>
        requestHumanReview(claim, approval.id),
      );
      return approval.promise;
    },
  },
});
// <end_wf>

restate.serve({ services: [agent, humanApprovalWorfklow] });

File: langchain-python/tour-of-agents/app/sub_workflow_agent.py
"""Workflow-as-tool: a tool that calls a separate Restate service which
contains the human-approval logic.

Putting the HITL flow in its own service lets you scale, audit, and version
it independently of the agent."""

import restate
from langchain.agents import create_agent
from langchain.chat_models import init_chat_model
from langchain_core.tools import tool

from restate.ext.langchain import RestateMiddleware, restate_context

from utils.models import ClaimPrompt, InsuranceClaim
from utils.utils import request_human_review


# <start_wf>
# Sub-workflow service for human approval.
human_approval_workflow = restate.Service("HumanApprovalWorkflow")


@human_approval_workflow.handler()
async def review(ctx: restate.Context, claim: InsuranceClaim) -> str:
    """Request human approval for a claim and wait for response."""
    approval_id, approval_promise = ctx.awakeable(type_hint=str)

    await ctx.run_typed(
        "Request review", request_human_review, claim=claim, awakeable_id=approval_id
    )

    return await approval_promise


# <end_wf>


# <start_here>
@tool
async def human_approval(claim: InsuranceClaim) -> str:
    """Ask for human approval for high-value claims."""
    return await restate_context().service_call(review, claim)


# <end_here>


agent = create_agent(
    model=init_chat_model("openai:gpt-4o-mini"),
    tools=[human_approval],
    system_prompt=(
        "You are an insurance claim evaluation agent. Use these rules:\n"
        "- if the amount is more than 1000, ask for human approval using tools;\n"
        "- if the amount is less than 1000, decide by yourself."
    ),
    middleware=[RestateMiddleware()],
)


agent_service = restate.Service("SubWorkflowClaimAgent")


@agent_service.handler()
async def run(_ctx: restate.Context, req: ClaimPrompt) -> str:
    result = await agent.ainvoke({"messages": [{"role": "user", "content": req.message}]})
    return result["messages"][-1].content


if __name__ == "__main__":
    import asyncio

    import hypercorn
    import hypercorn.asyncio

    app = restate.app(services=[agent_service, human_approval_workflow])
    conf = hypercorn.Config()
    conf.bind = ["0.0.0.0:9080"]
    asyncio.run(hypercorn.asyncio.serve(app, conf))
